### PR TITLE
fix(nn): add custom_jvp for log_softmax to prevent NaN gradients (fixes #40409)

### LINKED ISSUE #40409: jax.nn.log_softmax / jax.scipy.special.log_softmax gradient returns NaN on large float16 arrays despite finite forward pass
### Description

`jax.grad` through `log_softmax` on a sufficiently large (~2^16+ element) `float16` array returns an all-`NaN` gradient, even though the forward pass is finite and correct. Reproduces via both `jax.scipy.special.log_softmax` and `jax.nn.log_softmax`, in eager mode and under `jax.jit`.

The true gradient is small, not out of range: for `log_softmax(x).sum()` it is `1 - n * softmax(x)_k`, which for the repro below evaluates to about `[-13.66, 0.98]` (confirmed both via `float32` autodiff and the closed-form expression). `float16`'s max magnitude is 65504, so there's no numerical reason for this to be anything but a small finite value. A same-size, same-dtype *linear* reduction whose true gradient genuinely does exceed `float16`'s range (`x - jnp.sum(x)`) correctly saturates to `-inf` instead — so this isn't range saturation, it's an indeterminate form (`0 * inf` / `inf - inf`) surviving inside the VJP.

This looks related to #13529 at the implementation level (`log_softmax` subtracts a `stop_gradient`'d `jnp.max(x)` before exponentiating), but isn't the same trigger: that issue traces to gradient "smearing" across tied max elements on a deliberately degenerate all-zero-logit input. Forcing this repro's input to have a single, unambiguous, untied max still produces a full `NaN` gradient, and this uses ordinary `N(0,1)` data with no special construction.

## Minimal Reproducer

```python
import jax
import jax.numpy as jnp

x = jax.random.normal(jax.random.PRNGKey(0), (65536,), dtype=jnp.float16)

# Forward pass: correct and finite
y = jax.scipy.special.log_softmax(x)
print(jnp.isfinite(y).all())   # True

# Gradient: silently wrong
g = jax.grad(lambda x: jax.scipy.special.log_softmax(x).sum())(x)
print(jnp.isfinite(g).all())     # False
print((~jnp.isfinite(g)).sum())  # 65536 -- every element is NaN

# Also affects jax.nn.log_softmax, and JIT:
g2 = jax.jit(jax.grad(lambda x: jax.nn.log_softmax(x, axis=-1).sum()))(x)
print(jnp.isfinite(g2).all())  # False

# True gradient is small, well within float16's range:
g32 = jax.grad(lambda x: jax.scipy.special.log_softmax(x).sum())(x.astype(jnp.float32))
print(float(g32.min()), float(g32.max()))   # ~-13.66, ~0.98

# Contrast: a linear reduction whose true gradient IS out of range
# saturates to -inf, not NaN, as expected:
g3 = jax.grad(lambda x: (x - jnp.sum(x)).sum())(x)
print(jnp.isneginf(g3).all())  # True

# Ruling out max-tie smearing (cf. #13529): force a single, unambiguous max
x_unique_max = x.at[0].set(jnp.float16(6.0))
g4 = jax.grad(lambda x: jax.scipy.special.log_softmax(x).sum())(x_unique_max)
print(jnp.isfinite(g4).all())  # Still False
```

### Error

No exception is raised — NaN is returned silently.

### Expected Behavior

The gradient should be finite for a well-conditioned finite input where the true gradient is small, matching float32 behavior on the same data.

### System info (python version, jaxlib version, accelerator, etc.)

jax:    0.10.1
jaxlib: 0.10.1
numpy:  2.4.6
python: 3.12.13 | packaged by Anaconda, Inc. | (main, Mar 19 2026, 20:20:58) [GCC 14.3.0]
device info: cpu-1, 1 local devices
process_count: 1
platform: uname_result(system='Linux', release='6.8.0-136-generic', machine='x86_64')
JAX_PLATFORMS=cpu

### PR DESCRIPTION
**Pull request by jdymitarai**
Fixes #40409

### Motivation and Context
`jax.grad` through `log_softmax` on large arrays (~2^16+ elements, e.g. float16) previously produced `NaN` gradients due to intermediate exponentiation sums in the forward pass causing indeterminate forms (`0 * inf`) during reverse-mode autodiff.

As noted in `tests/nn_test.py:625`:
> `# TODO(mattjj): include log_softmax in these extra tests if/when we add a custom_jvp rule for it (since otherwise it doesn't pass the numerical checks below).`

### Description of Changes
1. Added `_log_softmax` with `@partial(custom_derivatives.custom_jvp, nondiff_argnums=(1,))` and its accompanying `_log_softmax_jvp` rule:
   $$\dot{y} = \dot{x} - \sum (\mathrm{softmax}(x) \cdot \dot{x})$$
   which is numerically stable and avoids intermediate `exp` gradient overflows.
2. Wired `log_softmax` to delegate to `_log_softmax` when `config.softmax_custom_jvp.value` is enabled, preserving backward compatibility via `_log_softmax_deprecated`.
3. Updated `jax.nn.softmax` and `jax.nn.log_softmax` signatures to accept optional `initial: ArrayLike = -np.inf`.
4. Resolved the TODO in `tests/nn_test.py` by enabling second-order gradient checks for `log_softmax` under `testSoftmaxWhereMask`, and added `testLogSoftmaxGrad`, `testLogSoftmaxGradResiduals`, and `testLogSoftmaxGradFlag`.

### CODE REVIEW

**Comment by jdymitarai** [comment]
Closing for now to keep the PR queue focused. Thanks!
